File: clients/ios/Lextures/Features/Paths/DashboardStudySection.swift
import SwiftUI

struct DashboardWhatsNext: Hashable {
    var course: CourseSummary
    var primary: LearnerRecommendationItem?
    var chips: [LearnerRecommendationItem]
    var degraded: Bool
}

struct DashboardStudySection: View {
    @Environment(AuthSession.self) private var session
    @Environment(\.colorScheme) private var colorScheme

    let studentCourses: [CourseSummary]
    let onOpenReview: () -> Void
    let onOpenRecommendation: (CourseSummary, CourseStructureItem) -> Void
    let onOpenPaths: () -> Void

    @State private var myPaths: [PathProgress] = []
    @State private var whatsNext: DashboardWhatsNext?
    @State private var loadingPaths = false
    @State private var loadingRecs = false

    var body: some View {
        Group {
            if !myPaths.isEmpty {
                pathsCard
            }
            if let whatsNext {
                recommendationsCard(whatsNext)
            }
        }
        .task(id: studentCourses.map(\.id).joined(separator: ",")) {
            await loadPaths()
            await loadRecommendations()
        }
    }

    private var pathsCard: some View {
        LMSCard {
            VStack(alignment: .leading, spacing: 10) {
                HStack {
                    LMSSectionHeader(title: L.text("mobile.paths.dashboardTitle"), systemImage: "point.topleft.down.to.point.bottomright.curvepath")
                    Spacer(minLength: 0)
                    Button(L.text("mobile.paths.viewAll"), action: onOpenPaths)
                        .font(.caption.weight(.semibold))
                }
                ForEach(myPaths.prefix(3)) { path in
                    VStack(alignment: .leading, spacing: 4) {
                        HStack {
                            Text(path.pathTitle)
                                .font(.caption.weight(.semibold))
                            Spacer(minLength: 0)
                            Text(path.progressLabel)
                                .font(.caption2)
                                .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
                        }
                        ProgressView(value: Double(path.percent), total: 100)
                            .tint(LexturesTheme.primary)
                    }
                }
            }
        }
    }

    @ViewBuilder
    private func recommendationsCard(_ bundle: DashboardWhatsNext) -> some View {
        LMSCard(accent: LexturesTheme.primary) {
            VStack(alignment: .leading, spacing: 8) {
                HStack(spacing: 6) {
                    Image(systemName: "sparkles")
                        .foregroundStyle(LexturesTheme.primary)
                    Text(bundle.course.displayTitle)
                        .font(.caption.weight(.semibold))
                        .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
                }
                if let primary = bundle.primary {
                    Text(primary.title)
                        .font(.subheadline.weight(.semibold))
                        .foregroundStyle(LexturesTheme.textPrimary(for: colorScheme))
                    Text(primary.reason)
                        .font(.caption)
                        .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
                    if bundle.degraded {
                        Text(L.text("mobile.paths.recommendationsDegraded"))
                            .font(.caption2)
                            .foregroundStyle(LexturesTheme.amber)
                    }
                    Button(L.text("mobile.paths.go")) {
                        if primary.itemType == "review_card" {
                            onOpenReview()
                        } else {
                            Task { await openRecommendation(course: bundle.course, item: primary) }
                        }
                        Task { await recordClick(course: bundle.course, item: primary, rank: 0) }
                    }
                    .font(.caption.weight(.semibold))
                    .buttonStyle(.borderedProminent)
                    .tint(LexturesTheme.primary)
                    if !bundle.chips.isEmpty {
                        ScrollView(.horizontal, showsIndicators: false) {
                            HStack(spacing: 8) {
                                ForEach(Array(bundle.chips.enumerated()), id: \.element.id) { index, chip in
                                    Button(chip.title) {
                                        if chip.itemType == "review_card" {
                                            onOpenReview()
                                        } else {
                                            Task { await openRecommendation(course: bundle.course, item: chip) }
                                        }
                                        Task { await recordClick(course: bundle.course, item: chip, rank: index + 1) }
                                    }
                                    .font(.caption2.weight(.semibold))
                                    .padding(.horizontal, 10)
                                    .padding(.vertical, 6)
                                    .background(LexturesTheme.primary.opacity(0.12))
                                    .clipShape(Capsule())
                                }
                            }
                        }
                    }
                } else {
                    Text(L.format("mobile.paths.caughtUpInCourse", bundle.course.displayTitle))
                        .font(.caption)
                        .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
    }

    private func loadPaths() async {
        guard let token = session.accessToken else { return }
        loadingPaths = true
        defer { loadingPaths = false }
        myPaths = (try? await LMSAPI.fetchMyPaths(accessToken: token)) ?? []
    }

    private func loadRecommendations() async {
        guard let token = session.accessToken,
              let userId = NotebookStore.jwtSubject(from: token),
              let course = studentCourses.first(where: \.viewerIsStudent) ?? studentCourses.first
        else {
            whatsNext = nil
            return
        }
        loadingRecs = true
        defer { loadingRecs = false }
        do {
            let responses = try await withThrowingTaskGroup(of: LearnerRecommendationsResponse.self) { group in
                for surface in PathsLogic.recommendationSurfaces {
                    group.addTask {
                        try await LMSAPI.fetchLearnerRecommendations(
                            userId: userId,
                            courseId: course.id,
                            surface: surface,
                            accessToken: token,
                            limit: 4
                        )
                    }
                }
                var out: [LearnerRecommendationsResponse] = []
                for try await response in group { out.append(response) }
                return out
            }
            let merged = PathsLogic.mergeRecommendations(responses)
            whatsNext = DashboardWhatsNext(
                course: course,
                primary: merged.primary,
                chips: merged.chips,
                degraded: merged.degraded
            )
            if let primary = merged.primary {
                try? await LMSAPI.postRecommendationEvent(
                    body: RecommendationEventBody(
                        courseId: course.id,
                        itemId: primary.itemId,
                        surface: primary.surface,
                        eventType: "impression",
                        rank: 0
                    ),
                    accessToken: token
                )
            }
        } catch {
            whatsNext = nil
        }
    }

    private func openRecommendation(course: CourseSummary, item: LearnerRecommendationItem) async {
        guard let token = session.accessToken else { return }
        let structure = (try? await LMSAPI.fetchCourseStructure(courseCode: course.courseCode, accessToken: token)) ?? []
        guard let target = PathsLogic.structureItem(for: item, in: structure) else { return }
        onOpenRecommendation(course, target)
    }

    private func recordClick(course: CourseSummary, item: LearnerRecommendationItem, rank: Int) async {
        guard let token = session.accessToken else { return }
        try? await LMSAPI.postRecommendationEvent(
            body: RecommendationEventBody(
                courseId: course.id,
                itemId: item.itemId,
                surface: item.surface,
                eventType: "click",
                rank: rank
            ),
            accessToken: token
        )
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/features/paths/DashboardStudySection.kt
package com.lextures.android.features.paths

import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material3.Button
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.lextures.android.R
import com.lextures.android.core.auth.AuthSession
import com.lextures.android.core.design.LexturesColors
import com.lextures.android.core.design.accentColor
import com.lextures.android.core.design.textPrimary
import com.lextures.android.core.design.textSecondary
import com.lextures.android.core.i18n.L
import com.lextures.android.core.i18n.LocalLocalePreferences
import com.lextures.android.core.lms.CourseStructureItem
import com.lextures.android.core.lms.CourseSummary
import com.lextures.android.core.lms.LearnerRecommendationItem
import com.lextures.android.core.lms.LmsApi
import com.lextures.android.core.lms.PathProgress
import com.lextures.android.core.lms.PathsLogic
import com.lextures.android.core.lms.RecommendationEventBody
import com.lextures.android.core.notebook.NotebookStore
import com.lextures.android.features.home.LmsCard
import com.lextures.android.features.home.LmsSectionHeader

data class DashboardWhatsNext(
    val course: CourseSummary,
    val primary: LearnerRecommendationItem?,
    val chips: List<LearnerRecommendationItem>,
    val degraded: Boolean,
)

@Composable
fun DashboardStudySection(
    session: AuthSession,
    studentCourses: List<CourseSummary>,
    learningPathsEnabled: Boolean,
    onOpenReview: () -> Unit,
    onOpenRecommendation: (CourseSummary, CourseStructureItem) -> Unit,
    onOpenPaths: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val accessToken by session.accessToken.collectAsState()
    val context = LocalContext.current
    val localePrefs = LocalLocalePreferences.current
    val scope = rememberCoroutineScope()

    var myPaths by remember { mutableStateOf<List<PathProgress>>(emptyList()) }
    var whatsNext by remember { mutableStateOf<DashboardWhatsNext?>(null) }

    LaunchedEffect(accessToken, studentCourses.map { it.id }.joinToString(), learningPathsEnabled) {
        val token = accessToken ?: return@LaunchedEffect
        if (learningPathsEnabled) {
            myPaths = runCatching { LmsApi.fetchMyPaths(token) }.getOrDefault(emptyList())
        } else {
            myPaths = emptyList()
        }

        val learnerId = NotebookStore.jwtSubject(token) ?: return@LaunchedEffect
        val course = studentCourses.firstOrNull { it.viewerIsStudent } ?: studentCourses.firstOrNull()
            ?: return@LaunchedEffect
        try {
            val responses = coroutineScope {
                PathsLogic.recommendationSurfaces.map { surface ->
                    async {
                        LmsApi.fetchLearnerRecommendations(learnerId, course.id, surface, token, limit = 4)
                    }
                }.awaitAll()
            }
            val merged = PathsLogic.mergeRecommendations(responses)
            whatsNext = DashboardWhatsNext(course, merged.primary, merged.chips, merged.degraded)
            merged.primary?.let { primary ->
                runCatching {
                    LmsApi.postRecommendationEvent(
                        RecommendationEventBody(
                            courseId = course.id,
                            itemId = primary.itemId,
                            surface = primary.surface,
                            eventType = "impression",
                            rank = 0,
                        ),
                        token,
                    )
                }
            }
        } catch (_: Exception) {
            whatsNext = null
        }
    }

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(10.dp)) {
        if (myPaths.isNotEmpty()) {
            LmsCard {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                        LmsSectionHeader(L.text(context, localePrefs, R.string.mobile_paths_dashboardTitle))
                        TextButton(onClick = onOpenPaths) {
                            Text(L.text(context, localePrefs, R.string.mobile_paths_viewAll))
                        }
                    }
                    myPaths.take(3).forEach { path ->
                        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                            Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                                Text(path.pathTitle, fontSize = 12.sp, fontWeight = FontWeight.SemiBold, color = textPrimary())
                                Text(path.progressLabel, fontSize = 11.sp, color = textSecondary())
                            }
                            LinearProgressIndicator(progress = { path.percent / 100f }, modifier = Modifier.fillMaxWidth())
                        }
                    }
                }
            }
        }

        whatsNext?.let { bundle ->
            LmsCard(accent = accentColor()) {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    Text(bundle.course.displayTitle, fontSize = 12.sp, fontWeight = FontWeight.SemiBold, color = textSecondary())
                    bundle.primary?.let { primary ->
                        Text(primary.title, fontSize = 15.sp, fontWeight = FontWeight.SemiBold, color = textPrimary())
                        Text(primary.reason, fontSize = 12.sp, color = textSecondary())
                        if (bundle.degraded) {
                            Text(
                                L.text(context, localePrefs, R.string.mobile_paths_recommendationsDegraded),
                                fontSize = 11.sp,
                                color = LexturesColors.Amber,
                            )
                        }
                        Button(
                            onClick = {
                                if (primary.itemType == "review_card") {
                                    onOpenReview()
                                } else {
                                    scope.launch {
                                        val token = accessToken ?: return@launch
                                        val structure = runCatching {
                                            LmsApi.fetchCourseStructure(bundle.course.courseCode, token)
                                        }.getOrDefault(emptyList())
                                        PathsLogic.structureItem(primary, structure)?.let {
                                            onOpenRecommendation(bundle.course, it)
                                        }
                                    }
                                }
                                scope.launch {
                                    val token = accessToken ?: return@launch
                                    runCatching {
                                        LmsApi.postRecommendationEvent(
                                            RecommendationEventBody(
                                                courseId = bundle.course.id,
                                                itemId = primary.itemId,
                                                surface = primary.surface,
                                                eventType = "click",
                                                rank = 0,
                                            ),
                                            token,
                                        )
                                    }
                                }
                            },
                        ) {
                            Text(L.text(context, localePrefs, R.string.mobile_paths_go))
                        }
                        if (bundle.chips.isNotEmpty()) {
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .horizontalScroll(rememberScrollState()),
                                horizontalArrangement = Arrangement.spacedBy(8.dp),
                            ) {
                                bundle.chips.forEachIndexed { index, chip ->
                                    TextButton(
                                        onClick = {
                                            if (chip.itemType == "review_card") {
                                                onOpenReview()
                                            } else {
                                                scope.launch {
                                                    val token = accessToken ?: return@launch
                                                    val structure = runCatching {
                                                        LmsApi.fetchCourseStructure(bundle.course.courseCode, token)
                                                    }.getOrDefault(emptyList())
                                                    PathsLogic.structureItem(chip, structure)?.let {
                                                        onOpenRecommendation(bundle.course, it)
                                                    }
                                                }
                                            }
                                            scope.launch {
                                                val token = accessToken ?: return@launch
                                                runCatching {
                                                    LmsApi.postRecommendationEvent(
                                                        RecommendationEventBody(
                                                            courseId = bundle.course.id,
                                                            itemId = chip.itemId,
                                                            surface = chip.surface,
                                                            eventType = "click",
                                                            rank = index + 1,
                                                        ),
                                                        token,
                                                    )
                                                }
                                            }
                                        },
                                    ) {
                                        Text(chip.title, fontSize = 11.sp)
                                    }
                                }
                            }
                        }
                    } ?: Text(
                        context.getString(R.string.mobile_paths_caughtUpInCourse, bundle.course.displayTitle),
                        fontSize = 12.sp,
                        color = textSecondary(),
                    )
                }
            }
        }
    }
}